qiskit/ignis/measurement/discriminator/filters.py

```diff
@@ -0,0 +1,78 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=cell-var-from-loop
+
+
+"""
+Discrimination filters.
+
+"""
+from copy import deepcopy
+
+from qiskit.exceptions import QiskitError
+from qiskit.ignis.characterization.fitters import BaseFitter
+from qiskit.result.result import Result
+from qiskit.result.models import ExperimentResultData
+
+
+class DiscriminationFilter:
```

Comment:
This is a personal opinion that goes against the Ignis convention, but I don't like the idea of the `Filter`. `Fitter` are already stateful as they contain the fitting data. As far as I know the only purpose a `Filter` serves is to produce a new discriminated measurement result. It seems that introducing a new class just for this is a burden on the user in terms of signal to noise ratio. For example, looking at the fitters in `scipy` they provide a single class with methods `fit` and `predict`. Thoughts @dcmckayibm?

Reply:
The idea is that a `Fitter` operates on calibration data to produce some type of characterization and a `Filter` is an object that transforms results into results. Not every `Fitter` has need for a `Filter` (e.g. T1), which is one rationale for a new class. Also once created the `Filter` can be passed around independently of the `Fitter`. This is already how it's done for the measurement correction (right or wrong).

Reply:
I understand this is how it is done currently 😄 and my above comment is non-blocking. The above is my off-topic opinion. 

Another option would be having the Filter be an object that the Fitter has a reference too and has a method `discriminate`. It could still be returned by the `fit` method. Since the `Filter` class is the same for different discriminator implementations, there is really no need for the user to every has to see or use it.

I just worry that sometimes from a user's perspective too many objects can be cumbersome and increase the signal to noise ratio when it comes to understanding the code. Consider how simple to use `sklearn.discriminant_analysis` is to use, which accomplishes the same task without a Filter object.

Reply:
Fitter operates on calibration data, Filter on external data. Do we *really* need it? That could be asked of any object-oriented class distinction.

Reply:
My preferred way of using the discriminator would be à la `scikit-learn`
```
discriminator = LinearIQDiscriminator(cal_results, expected_states, ...)  # fit() done in here
new_results = discriminator.predict(results)
```
If I understand the `fitter` / `filter` idea properly, the `filter` requires the user to type in an extra line of code:
```
discriminator = LinearIQDiscriminator(cal_results, expected_states, ...)  # fit() done in here
filter = DiscriminationFilter(discriminator)
new_results = filter.predict(results)
```
and when you look under the hood you realize that the `filter` was more or less the discriminator all along. I think introducing an extra class to handle external data, which is very much like calibration data but without the expected states, is unnecessary.